Refactoring around the `glassfish` module

dmatej opened the pull request:
* Removed unused code
* JDepend is happy - no cyclic dependencies between packages
* `static` renamed to `embedded`, however static had to remain in code for the embedded-glassfish-maven-plugin which uses hardcoded `static` value; it is marked as deprecated now.
* `*Static*` classes renamed to Embedded
* Anonymous or private classes moved to own files

### Tests
* Passed SpotBugs rules (without exclusions)
* TCK passed for glassfish which came out of build https://github.com/eclipse-ee4j/glassfish/issues/1:
  * https://ci.eclipse.org/jakartaee-tck/job/10/job/eftl-jakartaeetck-run-100/245/parameters/
  * https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck/job/10.0.x/143/

### Review 
* basically diff makes nearly no sense, rather open the implementation and try to find issues. It should be much easier than before.

dmatej committed:
SpotBugs: Explicit dependency on system charset or UTF-8

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Deleted unused Rejar class failing to pass SpotBugs rules

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using foreach instead of enumeration

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Refactoring: Cohesion, coupling, dependency cycles

- also removed some dead code and unused classes.
- bootstrap.Constants renamed to BootstrapKeys and moved to cfg package
- new osgi.impl package now contains classes which were inside MainHelper
- PlatformHelper.init body moved to constructor
- new log package to avoid cycles

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
GlassfishBundleContext wrapper to hide deprecated classes from OSGI

- Rewriting the code to avoid deprecated APIs is nontrivial and it seems also
  less effective. I have found some old discussion where people told that those
  deprecated APIs will not be removed any time soon (in 2009).
  Until somebody would have better knowledge to rewrite it correctly it seems
  we can continue with the deprecated code. At least, I moved it aside to own
  class.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
BundleProvisionerCustomizer classes moved to own files

- Also some minor improvements
- Cleaner filtering files (just syntax with side effects replaced)
- Removed using reflection to call constructors of known classes

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Constants merged into BootstrapKeys as both refer to the same cfg file

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The COMMAND_PATTERN was unused

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Deleted GlassFishRuntimeDecorator

- extended just once, did not make much sense

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Platform helpers renamed to OsgiPlatformadapters

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added explanation for FelixUtil

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Cleanup: Deleted Configurator, renamed variables, reduced visibility

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
AsenvConf has own class now

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Own GFBootstrapProperties for the GF bootstrap

- allows nicer initialization and validation

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Moving methods, making them private

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
GlassFishDecorator moved to osgi package

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Autodetection of the install root moved and bit improved

- slowly dismantling the MainHelper
- depends on the class which really needs it

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
AutoDisposableGlassFish class

- converted from anonymous class
- TODO: the autodelete feature should be more explicit than depending on
  instance root property set by the user.
- removal updated to NIO2, added logging errors

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
The getPropertyOrSystemProperty moved to its only usage

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Util.copy moved to its only usage, try-with, narrowed

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Util class dismantled and removed

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Launcher moved to own file

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Dismantled MainHelper

- Methods moved or deleted
- Created StartupContextCfgFactory
  - It is required by GlassFishMain, however the StartupContextCfg is used
    also from other places with no access to the factory.
  - Basically it is very similar to BootstrapProperties, but that is not visible
    to GlassFishMain.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
package.html transformed to package-info.java

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed UberJarOSGiGlassFishRuntimeBuilder logs declaring copy pasted class name

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Added FIXME comment about test class in production code

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Moving classes, Static->Embedded

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Simplification - removed need to create two OsgiPlatformAdapter instances

- One was used to read the second, originally it was pretty confusing
- However we still need to create StartupContextCfg twice

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Extracted ServerFiles

- instanceRoot and installRoot are now resolved just once, not saved to and
  loaded from properties again

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Classpath+classloader related classes moved to cp package

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/osgi/Jar.java:
```diff
@@ -85,47 +83,56 @@ public Jar(URI uri) {
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || !(obj instanceof Jar)) {
```
`obj == null` check redundant here

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/cfg/AsenvConf.java:
```diff
@@ -0,0 +1,106 @@
+
+    private final Properties properties;
+
+    public AsenvConf(File asenv, File configDir) {
```
Maybe `private` or package-private scope for ctor?

avpinchuk commented on nucleus/core/bootstrap/pom.xml:
```diff
@@ -129,6 +129,15 @@
                     </archive>
                 </configuration>
             </plugin>
```
Missing license header update )

avpinchuk commented on appserver/extras/embedded/common/bootstrap/src/main/java/org/glassfish/uberjar/bootstrap/UberJarOSGiGlassFishRuntimeBuilder.java:
```diff
@@ -161,24 +161,15 @@ public GlassFishRuntime build(BootstrapProperties bsOptions) throws GlassFishExc
 
-        System.setProperty(Constants.INSTALL_ROOT_PROP_NAME, installRoot);
-        System.setProperty(Constants.INSTANCE_ROOT_PROP_NAME, instanceRoot);
+        System.setProperty(BootstrapKeys.INSTALL_ROOT_PROP_NAME, installRoot);
```
Maybe better use static import for consistency with other code?

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/cfg/ServerFiles.java:
```diff
@@ -0,0 +1,88 @@
+    public File getOsgiHome(String envKey, String sysPropsKey, Path defaultSubdir) {
+        final String envProperty = System.getenv(envKey);
+        if (envProperty != null) {
+            return ServerFiles.toExistingFile(envProperty);
```
Redundant `ServerFiles` prefix

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/cfg/ServerFiles.java:
```diff
@@ -0,0 +1,88 @@
+        // try system property, which comes from asenv.conf
+        final String sysProperty = System.getProperty(sysPropsKey);
+        if (sysProperty != null) {
+            return ServerFiles.toExistingFile(sysProperty);
```
Here too

avpinchuk reviewed (commented):
This is a quick look ;)

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/GlassFishImpl.java:
```diff
@@ -30,31 +32,41 @@
-    private ServiceLocator habitat;
-    volatile Status status = Status.INIT;
+    private ServiceLocator serviceLocator;
+    private volatile Status status = Status.INIT;
```
Please move this initialization into ctor after `serviceLocator` assignment. This will guarantee thread visibility of the `serviceLocator` and `gfKernel` values.

dmatej replied:
I moved that to the constructor, but as the first line. It doesn't matter, if constructor fails, its fields will not be used.

avpinchuk replied:
In this case `start()` method may throw NPE if called from another thread. Field initialization done after `super()` but before fields assignment in constructor. We have next sequence: init `gfKernel` with default value (null) -> volatile write `state`-> `gfKernel` assignment -> volatile read `state` in `start()` method -> volatile write `state` -> plain read `gfKernel` (at the line 75). This assignment does not cover by the happens-before relation. And another thread can see null. Synchronized on `start` does not matters because `gfKernel` initialized without synchronization.

dmatej replied:
I am not sure if we understand each other, so maybe silly question - does it apply for the current code, when I moved it to the constructor? I believe then all three fields should be initialized in order.
I understand and absolutely agree with you about initialization outside constructor. Btw on other my projects tools prohibit calling nonstatic methods of the same instance from constructor - for similar reasons.

dmatej replied:
Is it possible to use an object with unfinished constructor in another thread? How could that happen? The only such scenario I can imagine is if we would start thread using `this` in constructor.
I really want to get it.

dmatej replied:
Aha, seems I forgot something important again ... you are right, thanks!
https://stackoverflow.com/a/10528614/375449

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/cfg/StartupContextUtil.java:
```diff
@@ -36,46 +39,70 @@ public final class StartupContextUtil {
+
+    private static File findBootstrapFile() {
+        try {
+            return Which.jarFile(StartupContextUtil.class);
```
In the original code used `ASMain.class`. Is not this a mistake or typo?

dmatej replied:
It uses any class from the jar file to locate the jar file, so I used the closest one instead of making dependency cycle between packages. I will add javadoc for sure.

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/embedded/EmbeddedGlassFishRuntimeBuilder.java:
```diff
@@ -42,12 +47,12 @@
+    private static final Logger LOG = LogFacade.BOOTSTRAP_LOGGER;
     private static final String JAR_EXT = ".jar";
-    final List<String> moduleExcludes = Arrays.asList("jsftemplating.jar", "gf-client-module.jar");
+    private static final List<String> MODULE_EXCLUDES = Arrays.asList("jsftemplating.jar", "gf-client-module.jar");
```
`List.of(...)`?

avpinchuk commented on nucleus/core/bootstrap/src/main/java/com/sun/enterprise/glassfish/bootstrap/embedded/JarUtil.java:
```diff
@@ -38,41 +40,41 @@
+    private static final String RAR_EXTENSION = ".rar";
 
-    public static final List<String> systemRarNames = Collections.unmodifiableList(
+    private static final List<String> SYSTEM_RAR_NAMES = Collections.unmodifiableList(
```
`List.of(...)` instead of `Collections.unmodifiableList(...)`?

avpinchuk commented:
LGTM except for the above shortcomings.

dmatej committed:
Resolved small mistakes based on the code review

- Added some javadocs
- Removed redundant code

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed thread visibility of GlassFishImpl

- The "being threadsafe" applies to the construction of an object as well,
  when setting any non-static, non-final, non-volatile field to a value other
  than its default value.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>
